### PR TITLE
feat(core): support block links on cmdk

### PR DESCRIPTION
**Pull request by fundon**
Upstreams: https://github.com/toeverything/blocksuite/pull/8260
Closes: [BS-1323](https://linear.app/affine-design/issue/BS-1323/粘贴-link-to-block-到-link-弹窗，不符合预期)

### CODE REVIEW

**Comment by fundon** [comment]
* **#8192** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/8192?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#8169** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/8169?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `canary`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @fundon and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/frontend/core/src/modules/quicksearch/impls/docs.ts [hunk 1]
```diff
@@ -57,23 +64,31 @@ export class DocsQuickSearchSession
...
+            }
+
+            const doc = docs.find(doc => doc.docId === resolvedDocId);
+
+            if (resolvedBlockId) {
+              if (doc && doc.blockId !== resolvedBlockId) {
+                doc.blockId = resolvedBlockId;
+                doc.blockContent = '';
+              } else {
+                docs.unshift({
```
Comment (pengx17) [inline packages/frontend/core/src/modules/quicksearch/impls/docs.ts]: better not use `unshift` (side effect) in `.map`

File: packages/frontend/core/src/components/blocksuite/block-suite-editor/specs/custom/spec-patchers.tsx [hunk 2]
```diff
@@ -322,11 +321,13 @@ export function patchQuickSearchService(framework: FrameworkProvider) {
...
               framework.get(CreationQuickSearchSession),
               (query: string) => {
-                if (
-                  (query.startsWith('http://') ||
-                    query.startsWith('https://')) &&
-                  resolveLinkToDoc(query) === null
-                ) {
+                const isLink =
+                  query.startsWith('http') &&
+                  (query.startsWith('://', 4) || query.startsWith('s://', 4));
```
Comment (pengx17) [inline packages/frontend/core/src/components/blocksuite/block-suite-editor/specs/custom/spec-patchers.tsx]: is this any better than startsWith('http://')?

File: packages/frontend/core/src/modules/quicksearch/impls/docs.ts [hunk 3]
```diff
@@ -57,23 +60,33 @@ export class DocsQuickSearchSession
...
+            if (!resolvedDocId) {
+              return docs;
+            }
+
+            const doc = docs.find(doc => doc.docId === resolvedDocId);
+
+            if (resolvedBlockId) {
+              if (doc && doc.blockId !== resolvedBlockId) {
+                doc.blockId = resolvedBlockId;
+                doc.blockContent = '';
```
Comment (pengx17) [inline packages/frontend/core/src/modules/quicksearch/impls/docs.ts]: please don't mutate input in `.map`

**Review by pengx17 — CHANGES_REQUESTED** [review]

File: packages/frontend/core/src/modules/quicksearch/impls/docs.ts [hunk 4]
```diff
@@ -57,23 +60,33 @@ export class DocsQuickSearchSession
       } else {
         const resolvedDoc = resolveLinkToDoc(query);
```
Comment (EYHN) [inline packages/frontend/core/src/modules/quicksearch/impls/docs.ts]: Can we separate this part from `DocsQuickSearchSession` and implement a separate `DocLinkQuickSearchSession`?

File: packages/frontend/core/src/components/blocksuite/block-suite-editor/specs/custom/spec-patchers.tsx (hunk 2 above)
Comment (fundon) [reply packages/frontend/core/src/components/blocksuite/block-suite-editor/specs/custom/spec-patchers.tsx]: No need to compare again from the starting position

Comment (pengx17) [reply packages/frontend/core/src/components/blocksuite/block-suite-editor/specs/custom/spec-patchers.tsx]: I think this is over optimization. You try to optimize one function call into two.

File: packages/frontend/core/src/modules/quicksearch/impls/docs.ts (hunk 4 above)
Comment (fundon) [reply packages/frontend/core/src/modules/quicksearch/impls/docs.ts]: Good idea. This is easier to maintain.

File: packages/frontend/core/src/modules/quicksearch/impls/docs.ts (hunk 3 above)
Comment (fundon) [reply packages/frontend/core/src/modules/quicksearch/impls/docs.ts]: The current search result may not match the `reslovedBlockId`.

Comment (pengx17) [reply packages/frontend/core/src/modules/quicksearch/impls/docs.ts]: I know, but it is hard to know here whether or not the doc is cached in the search result.
Mutating it is dangerous.

Try something like this:

```ts
const idx = docs.findIndex(doc => doc.docId === resolvedDocId);
const doc = { ...doc };
doc.blockId = xxx;
doc.blockContent = 'yyy';
return docs.toSpliced(idx, 1, doc)
```

Comment (fundon) [reply packages/frontend/core/src/modules/quicksearch/impls/docs.ts]: TIL `toSpliced`

File: packages/frontend/core/src/components/blocksuite/block-suite-editor/specs/custom/spec-patchers.tsx (hunk 2 above)
Comment (fundon) [reply packages/frontend/core/src/components/blocksuite/block-suite-editor/specs/custom/spec-patchers.tsx]: Ok, back

**Review by pengx17 — APPROVED** [review]

**Review by EYHN — APPROVED** [review]

**Pull request merged**